btag.h

```diff
@@ -0,0 +1,150 @@
+        };
+        return names.at(wp);
+    }
+    static WorkingPointsbTag getWPFromString(const std::string wp_string)
+    {
+        static const std::map< std::string,WorkingPointsbTag> wps = {
+            { "Tight", WorkingPointsbTag::Tight },
+            { "Medium", WorkingPointsbTag::Medium },
+            { "Loose", WorkingPointsbTag::Loose },
+        };
+        return wps.at(wp_string);
+    }
+    static const std::string& getScaleStr(UncScale scale, Correlation corr)
+    {
+        static const std::map<std::pair<UncScale, Correlation>, std::string> names = {
+            { {UncScale::Down, Correlation::nocorr}, "down" },
+            { {UncScale::Down, Correlation::uncorrelated}, "down_uncorrelated" },
+            { {UncScale::Down, Correlation::correlated}, "down_correlated" },
+            { {UncScale::Central, Correlation::nocorr}, "central"},
+            { {UncScale::Central, Correlation::uncorrelated}, "central"},
+            { {UncScale::Central, Correlation::correlated}, "central"},
+            { {UncScale::Up, Correlation::nocorr}, "up" },
+            { {UncScale::Up, Correlation::uncorrelated}, "up_uncorrelated" },
+            { {UncScale::Up, Correlation::correlated}, "up_correlated" },
+        };
+        return names.at(std::make_pair(scale, corr));
+    }
+
+    static bool sourceApplies(UncSource source, int Jet_Flavour)
+    {
+        if(source == UncSource::btagSFbc && (Jet_Flavour == 4 || Jet_Flavour==5) ) return true;
+        if(source == UncSource::btagSFlight && Jet_Flavour == 0 ) return true;
+        return false;
+    }
+
+    using histEffmap= std::map<std::pair<WorkingPointsbTag, int>, std::shared_ptr<TH2>> ;
+    bTagCorrProvider(const std::string& fileName, const std::string& efficiencyFileName) :
+        corrections_(CorrectionSet::from_file(fileName)),
+        deepJet_incl_(corrections_->at("deepJet_incl")),
+        deepJet_comb_(corrections_->at("deepJet_comb")),
+        deepJet_wp_values_(corrections_->at("deepJet_wp_values"))
+    {
+        if(efficiencyFileName.size()>0){
+            auto efficiencyFile = root_ext::OpenRootFile(efficiencyFileName);
+            const std::vector<std::string> WpNames = {"Loose", "Medium", "Tight"};
+            const std::vector<int> Flavours = {0, 4, 5};
+            for(const auto & flav : Flavours){
+                auto denum = root_ext::ReadCloneObject<TH2>(*efficiencyFile, "jet_pt_eta_"+std::to_string(flav), "", true);
+                for (const auto & WPName : WpNames){
+                    auto num = root_ext::ReadCloneObject<TH2>(*efficiencyFile,  "jet_pt_eta_"+std::to_string(flav)+"_"+WPName,"", true);
+                    num->Divide(denum);
+                    histMapEfficiency[std::make_pair(getWPFromString(WPName),flav)] = std::shared_ptr<TH2>(num);
+                }
+            }
+        }
+    }
+
+    float getWPvalue(WorkingPointsbTag wp) const {
+        return deepJet_wp_values_->evaluate({getWPStr(wp)});
+    }
```

Comment:
load wp map in the constructor -> converting to string each time and call evaluate is way to inefficient

Reply:
ok